Workflow sandbox (#164)
temporalio/sdk-python: README.md

Activities are just functions decorated with `@activity.defn`. Simply write different ones and pass those to the worker
to have different activities called during the test.

#### Workflow Sandbox

By default workflows are run in a sandbox to help avoid non-deterministic code. If a call that is known to be
non-deterministic is performed, an exception will be thrown in the workflow which will "fail the task" which means the
workflow will not progress until fixed.

> The sandbox is not foolproof and non-deterministic can still occur. It is simply a best-effort way to catch bad code

Author: ```suggestion
The sandbox is not foolproof and non-determinisms can still occur. It is simply a best-effort way to catch bad code
```

Reviewer: (no s at the end)

Author: Fixed

side-effect free on import and have their non-deterministic aspects restricted. By default the entire Python standard
library, `temporalio`, and a couple of other modules are passed through from outside of the sandbox. To update this
list, see "Customizing the Sandbox".

Restrictions preventing known non-deterministic library calls are achieved using proxy objects on modules wrapped around
the custom importer set in the sandbox. Many restrictions apply at workflow import time and workflow run time, while
some restrictions only apply at workflow run time. A default set of restrictions is included that prevents most
dangerous standard library calls. However it is known in Python that some otherwise-non-deterministic invocations, like
reading a file from disk via `open` or using `os.environ`, are done as part of importing modules. To customize what is
and isn't restricted, see "Customizing the Sandbox".

##### Avoiding the Sandbox

There are three increasingly-scoped ways to avoid the sandbox. Users are discouraged from avoiding the sandbox if
possible.

To remove restrictions around a particular block of code, use `with temporalio.workflow.unsafe.sandbox_unrestricted():`.

> The workflow will still be running in the sandbox, but no restrictions for invalid library calls will be applied.

Reviewer: That's nice

`temporalio.worker.workflow_sandbox.SandboxedWorkflowRunner()` so by changing it to the unsandboxed runner, the sandbox
will not be used at all.

##### Customizing the Sandbox

⚠️ WARNING: APIs in the `temporalio.worker.workflow_sandbox` module are not yet considered stable and may change in
future releases.

When creating the `Worker`, the `workflow_runner` is defaulted to
`temporalio.worker.workflow_sandbox.SandboxedWorkflowRunner()`. The `SandboxedWorkflowRunner`'s init accepts a
`restrictions` keyword argument that is defaulted to `SandboxRestrictions.default`. The `SandboxRestrictions` dataclass
is immutable and contains three fields that can be customized, but only two have notable value

###### Passthrough Modules

To make the sandbox quicker when importing known third party libraries, they can be added to the
`SandboxRestrictions.passthrough_modules` set like so:

```python
my_restrictions = dataclasses.replace(
    SandboxRestrictions.default,
    passthrough_modules=SandboxRestrictions.passthrough_modules_default | SandboxMatcher(access={"pydantic"}),
)
my_worker = Worker(..., runner=SandboxedWorkflowRunner(restrictions=my_restrictions))
```

> If an "access" match succeeds for an import, it will simpl;y be forwarded from outside of the sandbox. See the API for

Reviewer: ```suggestion
If an "access" match succeeds for an import, it will simply be forwarded from outside of the sandbox. See the API for
```

Author: Fixed

more details on exact fields and their meaning.

###### Invalid Module Members

> `SandboxRestrictions.invalid_module_members` contains a root matcher that applies to all module members. This is already

Reviewer 2: ```suggestion
`SandboxRestrictions.invalid_module_members` contains a root matcher that applies to all module members. This already
```

Author: Fixed

being used altogether:

```python
my_restrictions = dataclasses.replace(
    SandboxRestrictions.default,
    invalid_module_members=SandboxRestrictions.invalid_module_members_default | SandboxMatcher(
      children={"datetime": SandboxMatcher(use={"date"})},
    ),
)
my_worker = Worker(..., runner=SandboxedWorkflowRunner(restrictions=my_restrictions))
```

See the API for more details on exact fields and their meaning.

##### Known Sandbox Issues

Below are known sandbox issues. As the sandbox is developed and matures, some may be resolved.

###### Global Import/Builtins

Currently the sandbox still references/alters the global `sys.modules` and `builtins.__import__` fields while running
workflow code. In order to prevent affecting other sandboxed code a global lock is used to ensure other threads aren't
yielded to during sandbox run. Effort is ongoing to remove this global side effect and make `sys` and `builtins` truly
isolated, but in the meantime know that this can affect other imports occurring in the system. Users are encouraged to
not define workflows in Python applications with many other moving parts if possible, and to always test.

###### Sandbox is not Secure

> The sandbox is built to catch many non-determinisms and state sharing issues, but it is not secure. Some known bad calls

Reviewer: ```suggestion
The sandbox is built to catch many non-deterministic and state sharing issues, but it is not secure. Some known bad calls
```

Author: Fixed
